### ISSUE TITLE
Accuracy on training and validation sets different from sk-learn's

### ISSUE THREAD

**Issue by lorenzoFabbri** [skorch-dev/skorch#499]
I'm using Skorch to fit the following LSTM model:

```python
class Net(nn.Module):
    
    def __init__(self, input_size, hidden_size, num_layers, output_dim):
        super(Net, self).__init__()
        
        self.hidden_size = hidden_size
        self.num_layers = num_layers
        
        # Model
        self.model = nn.LSTM(input_size=input_size, hidden_size=hidden_size, 
                             num_layers=num_layers, batch_first=True, 
                             dropout=0)
        
        # FC layer(s)
        self.fc = nn.Linear(hidden_size, output_dim)
    
    def forward(self, x):
        
        # Initialization hidden state and cell state
        h0 = torch.zeros(self.num_layers, x.size(0), self.hidden_size).requires_grad_()
        c0 = torch.zeros(self.num_layers, x.size(0), self.hidden_size).requires_grad_()
        
        out, (hn, cn) = self.model(x, (h0.to(DEVICE), c0.to(DEVICE)))
        
        out = F.dropout(self.fc(out[:, -1, :]), p=0.7)
        
        return out
```

I'm using the following `NeuralNetClassifier`:

```python
# Callbacks
train_acc = EpochScoring(scoring='accuracy', on_train=True, 
                         name='train_acc', lower_is_better=False)
callbacks = [train_acc]

net = NeuralNetClassifier(
    # Features, hidden features, layers, output
    Net(MAX_LEN, 70, 1, NUM_CLASSES), 
    
    criterion=nn.CrossEntropyLoss, 
    
    optimizer=optim.Adam, 
    lr=0.001, 
    callbacks=callbacks, 
    
    batch_size=32, 
    max_epochs=30, 
    module_output_features=NUM_CLASSES, 
    device=DEVICE, 
    
    iterator_train__shuffle=True, 
    iterator_train__num_workers=3, 
    iterator_valid__shuffle=True, 
    iterator_valid__num_workers=3, 
    train_split=predefined_split(val_ds)
)

# Fit
net.fit(train_ds, y=train_target)
```

Everything works fine, and I get the following scores:

```
 epoch    train_acc    train_loss    valid_acc    valid_loss     dur
-------  -----------  ------------  -----------  ------------  ------
     20       0.4606        1.0110       0.4409        1.0406  2.3029
     21       0.4644        1.0026       0.4419        1.0394  2.2966
     22       0.4721        0.9921       0.4434        1.0457  2.2422
     23       0.4722        0.9920       0.4511        1.0414  2.2612
     24       0.4736        0.9850       0.4419        1.0405  2.2405
     25       0.4758        0.9785       0.4476        1.0445  2.2362
     26       0.4849        0.9687       0.4364        1.0653  2.2461
     27       0.4848        0.9689       0.4539        1.0399  2.2616
     28       0.4827        0.9607       0.4561        1.0337  2.2136
     29       0.4924        0.9545       0.4586        1.0313  2.2006
     30       0.4961        0.9511       0.4516        1.0345  2.3168
```

Since it's a multi-class classification task, I printed the classification report from sk-learn (on the same datasets as the ones used for fitting and computing the accuracy for validation).
On the training set, I get the following:
```
              precision    recall  f1-score   support

           0       0.38      0.19      0.25      6070
           1       0.26      0.17      0.20      3992
           2       0.37      0.65      0.47      5934

    accuracy                           0.35     15996
   macro avg       0.34      0.33      0.31     15996
weighted avg       0.35      0.35      0.32     15996
```

On the validation set, I get the following:
```
              precision    recall  f1-score   support

           0       0.37      0.19      0.25      1518
           1       0.27      0.18      0.21       998
           2       0.38      0.67      0.49      1483

    accuracy                           0.36      3999
   macro avg       0.34      0.34      0.32      3999
weighted avg       0.35      0.36      0.33      3999
```

I did not expect these results. The accuracies are both different from what Skorch provides.
Since I provide the split for computing the accuracy on the validation set, it does not do CV and I expected the accuracies to match. Am I missing something?

**Comment by BenjaminBossan** [comment]
The main reason seems to be this: `iterator_valid__shuffle=True`. Basically, you are shuffling your predictions, so I would expect the scores to be at chance level. If, after changing this, you still don't see an improvement:

* Are the scores seen in the report the same as you would expect from random predictions or are they better?
* Are you using the same train/valid split in the report as during training?
* From what I see, your `predict_proba` doesn't return probabilities. Not sure if this interferes with the report. Could you try passing the predictions through softmax?

Some other issues I see that are not related to your original issue:

* You are using dropout as the last operation, are you sure that this is good?
* You are initializing h0 and c0 during each foward pass (with requires grad), this is very inefficient.

**Comment by lorenzoFabbri** [comment]
I do not understand: if I just compute the predictions, it should always give the same results. I mean, I'm not training the model.

- They're actually worse.
- Yes, the same.
- I do not understand, sorry. I do not have any such method. I did not put a softmax layer since I'm using cross-entropy.

- Obviously, not! I forgot to remove it once I deleted the real last layer.
- I just followed an example I found online, and since the dataset is small, I did not care. At least for the moment.

**Comment by ottonemo** [comment]
What @BenjaminBossan means with shuffling is that you are enabling shuffling on the validation data loader which is used for predictions as well. This in turn means that all data that when you run `y_pred = model.predict(X_valid)` then `X_valid` is shuffled internally, which means that `y_pred` is not ordered in the same way that `X_valid` is. This in turn means that if you compare `y_pred` with `y_valid` then the pairs do not match anymore!

Set `iterator_valid__shuffle=False` to solve your problem.

**Comment by BenjaminBossan** [comment]
@lorenzoFabbri Any progress on this?
